### PR TITLE
Add new field "entityRecipientsForInformation" (#4153)

### PR DESCRIPTION
**Pull request by vlo-rte**
- In release note :
  -  In chapter :  Features
  -  Text : #4153 : Add new field "entityRecipientsForInformation"

### CODE REVIEW

File: services/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java [hunk 1]
```diff
@@ -56,6 +56,12 @@ public class UserCardData implements UserCard {
...
     private Boolean recipientVisible;
 
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    private Boolean recipientForInformationVisible;
+
     @JsonInclude(JsonInclude.Include.NON_EMPTY)
     private List<Recipient> recipientList;
+
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    private List<Recipient> recipientForInformationList;
```
Comment (freddidierRTE) [inline services/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java:66]: not needed

File: services/businessconfig/src/main/modeling/swagger.yaml [hunk 2]
```diff
@@ -690,10 +690,16 @@ definitions:
...
+      recipientForInformationVisible: # tells whether to display recipient for information dropdown when creating a user card
+        type: boolean
       recipientList:
         type: array
         items:
           $ref: '#/definitions/Recipient'
+      recipientForInformationList:
+        type: array
+        items:
+          $ref: '#/definitions/Recipient'
```
Comment (freddidierRTE) [inline services/businessconfig/src/main/modeling/swagger.yaml:702]: not needed

File: services/cards-publication/src/main/java/org/opfab/cards/publication/model/CardOperationData.java [hunk 3]
```diff
@@ -41,6 +41,10 @@ public class CardOperationData implements CardOperation {
     @JsonInclude(JsonInclude.Include.NON_EMPTY)
     private List<String> entityRecipientsIds;
 
+    @Singular
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    private List<String> entityRecipientsForInformationIds;
+
```
Comment (freddidierRTE) [inline services/cards-publication/src/main/java/org/opfab/cards/publication/model/CardOperationData.java:47]: not needed

**Comment by freddidierRTE** [comment]
If there is only recipient for information  , the acknowledged word shall not be visible
image

**Review by freddidierRTE — CHANGES_REQUESTED** [review]

**Comment by freddidierRTE** [comment]
Add documentation for method  usercardTemplateGateway.getSelectedEntityRecipientsForInformation

like in https://opfab.github.io/documentation/current/reference_doc/#_get_selected_recipient_entities_list

File: services/cards-consultation/src/main/java/org/opfab/cards/consultation/model/LightCardConsultationData.java [hunk 4]
```diff
@@ -159,6 +164,8 @@ public static LightCardConsultationData copy(Card other) {
             builder.entitiesAcks(other.getEntitiesAcks());
         if (other.getEntityRecipients() != null && ! other.getEntityRecipients().isEmpty())
             builder.entityRecipients(other.getEntityRecipients());
+        if (other.getEntityRecipientsForInformation() != null && ! other.getEntityRecipientsForInformation().isEmpty())
+            builder.entityRecipientsForInformation(other.getEntityRecipientsForInformation());
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-consultation/src/main/java/org/opfab/cards/consultation/model/LightCardConsultationData.java:168]: Done.

File: services/cards-publication/src/main/java/org/opfab/cards/publication/model/ArchivedCardPublicationData.java [hunk 5]
```diff
@@ -92,6 +92,9 @@ public class ArchivedCardPublicationData implements Card {
     @Indexed
     private List<String> entityRecipients;
 
+    @Indexed
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/model/ArchivedCardPublicationData.java:95]: Done.

File: services/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java (hunk 1 above)
Comment (vlo-rte) [reply services/businessconfig/src/main/java/org/opfab/businessconfig/model/UserCardData.java:66]: Done.

File: services/businessconfig/src/main/modeling/swagger.yaml (hunk 2 above)
Comment (vlo-rte) [reply services/businessconfig/src/main/modeling/swagger.yaml:702]: Done.

File: services/cards-publication/src/main/java/org/opfab/cards/publication/model/CardOperationData.java (hunk 3 above)
Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/model/CardOperationData.java:47]: Done.

File: services/cards-publication/src/main/modeling/swagger.yaml [hunk 6]
```diff
@@ -359,6 +359,15 @@ definitions:
         items:
           type: string
         example: ["Dispatcher","Planner"]
+      entityRecipientsForInformation:
+        description: >-
+          List of entity recipients for information. The difference between this field and field 'entityRecipients' is the 
+          sender of the card who will not see the entity ack if the entity recipient is in 'entityRecipientsForInformation' 
+          and not in 'entityRecipients'.
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-publication/src/main/modeling/swagger.yaml:366]: Done.

File: services/cards-publication/src/main/modeling/swagger.yaml [hunk 7]
```diff
@@ -572,6 +581,14 @@ definitions:
         description: List of entity recipients
         items:
           type: string
+      entityRecipientsForInformation:
+        description: >-
+          List of entity recipients for information. The difference between this field and field 'entityRecipients' is the
+          sender of the card who will not see the entity ack if the entity recipient is in 'entityRecipientsForInformation' 
+          and not in 'entityRecipients'.
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-publication/src/main/modeling/swagger.yaml:588]: Done.

File: services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardNotificationService.java [hunk 8]
```diff
@@ -74,6 +74,11 @@ public void notifyOneCard(CardPublicationData card, CardOperationTypeEnum type)
             card.getEntityRecipients().forEach(listOfEntityRecipients::add);
         cardOperation.setEntityRecipientsIds(listOfEntityRecipients);
 
+        List<String> listOfEntityRecipientsForInformation = new ArrayList<>();
+        if (card.getEntityRecipientsForInformation() != null)
+            card.getEntityRecipientsForInformation().forEach(listOfEntityRecipientsForInformation::add);
+        cardOperation.setEntityRecipientsForInformationIds(listOfEntityRecipientsForInformation);
+
```
[the remark this answers is not in the document]
Comment (vlo-rte) [reply services/cards-publication/src/main/java/org/opfab/cards/publication/services/CardNotificationService.java:81]: Done.

**Comment by vlo-rte** [comment]
> If there is only recipient for information , the acknowledged word shall not be visible image

Done.

**Comment by vlo-rte** [comment]
> Add documentation for method usercardTemplateGateway.getSelectedEntityRecipientsForInformation
> 
> like in https://opfab.github.io/documentation/current/reference_doc/#_get_selected_recipient_entities_list

Done.

**Comment by vlo-rte** [comment]
I've merged my pull request because the review has been done and now the builds are successful.

**Pull request merged**